x-pack/plugins/maps/public/classes/sources/table_source/table_source.ts

```diff
@@ -0,0 +1,217 @@
+  static type = SOURCE_TYPES.TABLE_SOURCE;
+
+  static createDescriptor(descriptor: Partial<TableSourceDescriptor>): TableSourceDescriptor {
+    return {
+      type: SOURCE_TYPES.TABLE_SOURCE,
+      __rows: descriptor.__rows || [],
+      columns: descriptor.columns || [],
+      term: descriptor.term || '',
+      id: descriptor.id || uuid(),
+    };
+  }
+
+  readonly _descriptor: TableSourceDescriptor;
+
+  constructor(descriptor: TableSourceDescriptor, inspectorAdapters?: Adapters) {
+    const sourceDescriptor = TableSource.createDescriptor(descriptor);
+    super(descriptor, inspectorAdapters);
+    this._descriptor = sourceDescriptor;
+  }
+
+  async getDisplayName(): Promise<string> {
+    // no need to localize. this is never rendered.
+    return `table source ${uuid()}`;
+  }
+
+  getSyncMeta(): VectorSourceSyncMeta | null {
+    return null;
+  }
+
+  async getPropertiesMap(
+    searchFilters: VectorJoinSourceRequestMeta,
+    leftSourceName: string,
+    leftFieldName: string,
+    registerCancelCallback: (callback: () => void) => void
+  ): Promise<PropertiesMap> {
+    const propertiesMap: PropertiesMap = new Map<string, BucketProperties>();
+
+    for (let i = 0; i < this._descriptor.__rows.length; i++) {
+      const bucketProperties: BucketProperties = {};
+      const row: TableSourceValue[] = this._descriptor.__rows[i];
+      let propKey: string | number | undefined;
+      for (let j = 0; j < row.length; j++) {
+        if (row[j].key === this._descriptor.term) {
+          propKey = row[j].value;
+        } else {
+          bucketProperties[row[j].key] = row[j].value;
+        }
+      }
+      if (propKey) {
+        propertiesMap.set(propKey.toString(), bucketProperties);
+      }
+    }
+
+    return propertiesMap;
+  }
+
+  getTermField(): IField {
+    const column = this._descriptor.columns.find((c) => {
+      return c.name === this._descriptor.term;
+    });
+
+    if (!column) {
+      throw new Error(
+        `Cannot find column for ${this._descriptor.term} in ${JSON.stringify(
+          this._descriptor.columns
+        )}`
+      );
+    }
+
+    return new TableField({
+      fieldName: column.name,
+      source: this,
+      origin: FIELD_ORIGIN.SOURCE,
+      dataType: column.type,
+    });
+  }
+
+  getWhereQuery(): Query | undefined {
+    return undefined;
+  }
+
+  hasCompleteConfig(): boolean {
+    return true;
+  }
+
+  getId(): string {
+    return this._descriptor.id;
+  }
+
+  getRightFields(): IField[] {
+    return this._descriptor.columns.map((column) => {
+      return new TableField({
+        fieldName: column.name,
+        source: this,
+        origin: FIELD_ORIGIN.SOURCE,
+        dataType: column.type,
+      });
+    });
+  }
+
+  getFieldNames(): string[] {
+    return this._descriptor.columns.map((column) => {
+      return column.name;
+    });
+  }
+
+  hasMatchingMetricField(fieldName: string): boolean {
```

Comment:
Should `hasMatchingMetricField` be renamed? You have done a good job of removing `metric` from the interface since for this source its not an aggregation. How about renaming to just `hasField`?

Also, `hasMatchingMetricField` is part of `IESAggSource` but its only used by ITermJoinSource. Can you remove `hasMatchingMetricField` from `IESAggSource`?

Reply:
Actually, how about just removing the method all together to keep the interface cleaner and just inlining `!!joinSource.getFieldByName(fieldName)` where used?